src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs

```diff
@@ -432,7 +432,7 @@ public Activity AddBaggage(string key, string? value)
         {
             KeyValuePair<string, string?> kvp = new KeyValuePair<string, string?>(key, value);
 
-            if (_baggage != null || Interlocked.CompareExchange(ref _baggage, new LinkedList<KeyValuePair<string, string?>>(kvp), null) != null)
+            if (_baggage != null || Interlocked.CompareExchange(ref _baggage, new LinkedList<KeyValuePair<string, string?>>(kvp, reversed: true), null) != null)
```

Comment:
Thank youf again for working on this. 
I'm sorry for adding even more confusion, but I don't think we should fix it like this.
Please correct me if I'm wrong, but the most common scenario during the transfer of baggage is something like this:
```CSharp
foreach (var item in baggage)
{
    activity.AddBaggage(item.Name, item.Value);
}
``` 
Here a sample of how Kestrel doing this: https://github.com/dotnet/aspnetcore/blob/c97a0020d8bab6d895bf821f6e47ee8722aa17d5/src/Hosting/Hosting/src/Internal/HostingApplicationDiagnostics.cs#L276

This means that with this change if before the transfer we had baggage like this
```
Key1=NewValue;Key1=OldValue
```
After transfer that would be:
```
Key1=OldValue;Key1=NewValue
```
So the result of `GetBaggageItem` will be changing on each transfer and would be totally unpredictable (Or we would need to require reversal of the baggage collection before adding that is very unintuitive)

This is what actually happening now in net core 3.1 and at lease from my point of view it's better to have a change in behavior then changes in order between transfers.

Alternatively, you could change `GetBaggageItem` method itself to return the last item with that key in the list, but that would mean fully iterating over the list each time (or making list double-linked), so I'm not sure if it worth it.

I do know how complicated is the process of exposing new public method is for your team, but in terms of our use case having something like `SetBaggageItem()` or `RemoveBaggageItem()` would solved all of the problems and it would've allowed to avoid all of the problematic cases discassed here. I don't think somebody actually want's two items with the same key (or if they do they could work with `Baggage` enumeration), but it's quite posible and easy to call `Add` twice in diffirent places or have the need to update value. 

Having the ability to update the list, you could always suggest people use it instead of `Add` when they are facing simular problems and it will prevernt inceasing baggage size when it's not needed.

Reply:
@AndreyTretyak This change is keeping the compatibility with the previous versions. in other word, we'll be consistent across the board. now, the callers of the Baggage APIs will have a consistent way to depend on regardless of which version of `System.Diagnostics.DiagnosticSource` package they will use. Changing the order is going to break someone else for sure. Imagine someone already working around the 3.1 behavior by storing the Baggages in the Activity object in reversed order as receive it from the deserialization. If we we change the order now, will break such guy and also add more risk breaking the scenario of using different versions of `System.Diagnostics.DiagnosticSource`. 

It is up to the consumer to handle the Baggage list as long as the order is well defined and maintained across versions. Kestrel can consume the list in reverse way if they like, you may request them to fix that.

We are open to add more APIs e.g. `SetBaggage` in the next release but for sure this will not be in 5.0.

Reply:
Thank you for the detailed explanation, I totally understand the need for maintaining behavior. Sorry for questioning it too much it's definitely too late for changing order, I liked new behaivour too much :)

What's needed from me in order for you to consider adding `SetBaggage` in the future versions? Should I create a separate API suggestion?

Reply:
> What's needed from me in order for you to consider adding SetBaggage in the future versions? Should I create a separate API suggestion?

Please log a new issue for this request and add your proposal to the description. please link the new issue to #42596 to have more context. Please follow the instructions in https://github.com/dotnet/runtime/blob/master/docs/project/api-review-process.md when filing the new issue. Thanks for all your feedback. it was really super helpful.